### DISCUSSION TITLE
How to change Copilot autocomplete key from Tab? [Visual Studio]

### DISCUSSION THREAD

**Discussion by TuTAH1** [community/community#65015] [Copilot Conversations] [answered]
### Select Topic Area

Question

### Body

Currently, it uses "Tab" for generating code, what is very unconvinient!

- When I want to use IDE's autocomplete, I can't do it via Tab, becouse Copilot already uses this binding
- When I want just to add tab, the Copilot suggestion sudennly occurs and I have ot wait it adds the code, Ctrl+Z, wait for IDE unfreezes. It's just a few second, but occurs constantly and knocks off the flow

I didn't find a binding for generating code in Settings → Envoriment → Keyboard. 
Settings → Github → Copilot doesn't have such option too.

Where to change it?

**Comment by arafajs** [comment]
1. Open Visual Studio and go to Tools > Options.
2. In the Options dialog, go to the Environment > Keyboard tab.
3. In the Show commands containing: field, type "Copilot".
4. The Edit.Copilot.TriggerInlineSuggestion command should appear in the list.
5. Click the pencil icon next to the command.
6. In the Press shortcut keys: field, type the new keyboard shortcut you want to use for Copilot autocomplete.
7. Click Assign.

**Reply by TuTAH1** [reply]
I can even delete any bindings of `Options →  Envoriment → Keyboard → Edit.Copilot.TriggerInlineSuggestion`,
![image](https://github.com/community/community/assets/15982179/19c918bb-9bad-42f8-b2ee-adff62440b9e)
 but Copilot still generates suggested code when I click Tab

**Reply by arafajs** [reply]
I think this is a known issue with Copilot and Visual Studio.
You can wait for a fix from GitHub. They are aware of this issue and are working on a solution.

**Reply by TuTAH1** [reply]
> I think this is a known issue with Copilot and Visual Studio. You can wait for a fix from GitHub. They are aware of this issue and are working on a solution.

Can you please give a link to this issue? I didn't found such issue [here](https://github.com/orgs/community/discussions/categories/copilot)

**Reply by arafajs** [reply]
https://learn.microsoft.com/en-us/answers/questions/1184163/github-copilot-doesnt-get-triggered-by-tab-it-call
https://github.com/orgs/community/discussions/7036

**Reply by TuTAH1** [reply]
These issues have simmilar, but not same problem. For me, Tab is succesfully triggers Copilot but I don't want it be.
In the [first](https://learn.microsoft.com/en-us/answers/questions/1184163/github-copilot-doesnt-get-triggered-by-tab-it-call) issue, it doesn't triggers by Tab, what now is fixed
In the [second ](https://github.com/orgs/community/discussions/7036)issue is about VSCode that seems to have an option for changing Copilot's autocomplete shortcut.

Are they aware of Copilot triggering by Tab while having no such keybinding for Copilot in Keyboard settings is not the expected behaviour?

**Reply by arafajs** [reply]
I found the same problem on Stackoverflow [here](https://stackoverflow.com/questions/76486363/how-do-i-change-the-auto-complete-hotkey-for-github-copilot-suggestion-in-visual).  and there is no answer too. 🫤

**Reply by TuTAH1** [reply]
> I found the same problem on Stackoverflow [here](https://stackoverflow.com/questions/76486363/how-do-i-change-the-auto-complete-hotkey-for-github-copilot-suggestion-in-visual). and there is no answer too. 🫤

There's [more relevant question](https://stackoverflow.com/questions/76234921) on Stackoverflow (unanswered too). But I don't think Github team monitoring questions there. 

I see currently there's no solution of this problem, should I consider such Copilot behaviour (triggering on Tab if there's no such binding in Keyboard settings) as a bug and create a separate disscussion of it with a "bug" tag?

**Reply by arafajs** [reply]
Yes. good idea.

**Reply by TuTAH1** [reply]
Althrough, the `Edit.Copilot.TriggerInlineSuggestion` if believe its name, should trigger (showing) a suggestion, not writting it. So it's not a bug, but a lack of a feature. I'll still create a seperate disscussion of adding several shortcuts to Keyboard settings, since even if this topic would get an workaround-answer (like editing Copilot extension's files), the related feature-request won't become obsolete.

**Comment by fbdesignpro** [comment]
This is preventing me from using copilot... I support all discussions related to this! Feels like this should be an easy one to implement, too.

**Reply by tim-hub** [reply]
For me, it prevents me to use VSCode.

- Github Copilot Inline suggestion on JetBrains, is OK to be rebind.
- When use VSCode, I cannot.


I really want to use VSCode more to save some money

**Comment by TwistedMinda** [comment]
<img width="550" alt="Screenshot 2023-10-26 at 22 07 37" src="https://github.com/community/community/assets/11213609/7c276392-13a6-4cec-bb62-5edbd6bb8dd3">

It was this garbage "Inline Chat my ass" command that I had to bind to something else

**Comment by Gabriel-Damas** [comment]
1. Abra o VSCode: Certifique-se de que o Visual Studio Code esteja aberto. 2. Acesse as configurações

**Comment by pri-shav** [comment]
I get that the Tab key in Copilot is causing issues with your IDE's autocomplete. To change it:

- Go to your IDE settings (look for "Environment" or "Editor").
- Find the Copilot or code generation action.
- Reassign a new key that won't clash with autocomplete.
- Save, restart your IDE, and you're good.

**You can also check this link:** [https://stackoverflow.com/questions/76533712/change-the-key-mapping-of-github-copilot-autocomplete-in-visual-studio](https://stackoverflow.com/questions/76533712/change-the-key-mapping-of-github-copilot-)[snowflake course](https://www.igmguru.com/oracle-training/snowflake-training/)[autocomplete-in-visual-studio](https://stackoverflow.com/questions/76533712/change-the-key-mapping-of-github-copilot-autocomplete-in-visual-studio)

**Reply by ayim** [reply, marked as answer]
The other option you have available is in Tools > Options > IntelliCode > Advanced > Accept code suggestion on right arrow.
Accepting the editor's suggestion should work together with Copilot unless you're using a third party extension like Resharper that overrides the completions list 

https://devblogs.microsoft.com/visualstudio/github-copilot-visual-studio-intellisense/

**Comment by octoxan** [comment]
Is there still no way to have a different keyboard shortcut for Copilot inline completions and the built in VS Code inline completions?

**Reply by Rmattgraham** [reply]
Finally found the solution: [here](https://stackoverflow.com/questions/73705091/accept-github-copilot-inline-suggestion-with-right-arrow-key-instead-of-tab-in-v)

**Reply by octoxan** [reply]
Yessss that actually did it. Bound mine to ` so it matches zshrc auto suggest. Tab for built in VS Code, ` for Copilot. Perfect! I hated trying to just tab through snippets I've set up in vs code and having to dismiss a billion nonsense copilot suggestions for tab to work lol

**Reply by ayim** [reply]
Going to close this issue since VSCode is unrelated - feel free to reopen a new issue for VSCode
